Data/Primitive/PrimArray.hs

```diff
@@ -0,0 +1,488 @@
+unsafeThawPrimArray (PrimArray arr#)
+  = primitive (\s# -> (# s#, MutablePrimArray (unsafeCoerce# arr#) #))
+
+-- | Read a primitive value from the array.
+indexPrimArray :: forall a. Prim a => PrimArray a -> Int -> a
+{-# INLINE indexPrimArray #-}
+indexPrimArray (PrimArray arr#) (I# i#) = indexByteArray# arr# i#
+
+sizeofPrimArray :: forall a. Prim a => PrimArray a -> Int
+{-# INLINE sizeofPrimArray #-}
+sizeofPrimArray (PrimArray arr#) = I# (quotInt# (sizeofByteArray# arr#) (sizeOf# (undefined :: a)))
+
+-- | Lazy right-associated fold over the elements of a 'PrimArray'.
+{-# INLINE foldrPrimArray #-}
+foldrPrimArray :: forall a b. Prim a => (a -> b -> b) -> b -> PrimArray a -> b
+foldrPrimArray f z arr = go 0
+  where
+    !sz = sizeofPrimArray arr
+    go !i
+      | sz > i = f (indexPrimArray arr i) (go (i+1))
+      | otherwise = z
+
+-- | Strict right-associated fold over the elements of a 'PrimArray'.
+{-# INLINE foldrPrimArray' #-}
+foldrPrimArray' :: forall a b. Prim a => (a -> b -> b) -> b -> PrimArray a -> b
+foldrPrimArray' f z0 arr = go (sizeofPrimArray arr - 1) z0
+  where
+    go !i !acc
+      | i < 0 = acc
+      | otherwise = go (i - 1) (f (indexPrimArray arr i) acc)
+
+-- | Lazy left-associated fold over the elements of a 'PrimArray'.
+{-# INLINE foldlPrimArray #-}
+foldlPrimArray :: forall a b. Prim a => (b -> a -> b) -> b -> PrimArray a -> b
+foldlPrimArray f z arr = go (sizeofPrimArray arr - 1)
+  where
+    go !i
+      | i < 0 = z
+      | otherwise = f (go (i - 1)) (indexPrimArray arr i)
+
+-- | Strict left-associated fold over the elements of a 'PrimArray'.
+{-# INLINE foldlPrimArray' #-}
+foldlPrimArray' :: forall a b. Prim a => (b -> a -> b) -> b -> PrimArray a -> b
+foldlPrimArray' f z0 arr = go 0 z0
+  where
+    !sz = sizeofPrimArray arr
+    go !i !acc
+      | i < sz = go (i + 1) (f acc (indexPrimArray arr i))
+      | otherwise = acc
+
+-- | Strict left-associated fold over the elements of a 'PrimArray'.
+{-# INLINE foldlPrimArrayM' #-}
+foldlPrimArrayM' :: (Prim a, Monad m) => (b -> a -> m b) -> b -> PrimArray a -> m b
+foldlPrimArrayM' f z0 arr = go 0 z0
+  where
+    !sz = sizeofPrimArray arr
+    go !i !acc1
+      | i < sz = do
+          acc2 <- f acc1 (indexPrimArray arr i)
+          go (i + 1) acc2
+      | otherwise = return acc1
+
+-- | Monadic map over a 'PrimArray'.
+{-# INLINE mapPrimArrayM #-}
+mapPrimArrayM :: (Monad m, Prim a, Prim b)
+  => (a -> m b)
+  -> PrimArray a
+  -> m (PrimArray b)
+mapPrimArrayM f arr = go 0 []
+  where
+  !sz = sizeofPrimArray arr
+  go !ix !xs
+    | ix < sz = do
+        !x <- f (indexPrimArray arr ix)
+        go (ix + 1) (x : xs)
+    | otherwise = return (primArrayFromBackwardsListN sz xs)
+
+-- | Monadic map over a 'PrimArray' where the monad allows
+--   primitive actions.
+{-# INLINE mapPrimArrayP #-}
+mapPrimArrayP :: (PrimBase m, Prim a, Prim b)
+  => (a -> m b)
+  -> PrimArray a
+  -> m (PrimArray b)
+mapPrimArrayP = unsafeMapPrimArrayP
```

Comment:
I thought about trying for an appropriate constraint for the polymorphic arrays, but decided to just export the unsafe one. Whatever we do, we should do consistently. I believe the correct constraint is probably a type family-based one rather than a class-based one. We don't actually need to build a dictionary to represent the validity evidence; we just need to know that it's valid.

```haskell
class AffineEnoughF m => AffineEnough m
instance AffineEnoughF m => AffineEnough m

type family AffineEnoughF m :: ()
type instance AffineEnoughF (ST s) = '()
type instance AffineEnoughF IO = '()
type instance AffineEnoughF (ExceptT e m) = AffineEnoughF m
type instance AffineEnoughF (StateT m) = AffineEnoughF m
type instance AffineEnoughF (ReaderT e m) = AffineEnoughF m
type instance AffineEnoughF (ListT m) = TypeError ...
type instance AffineEnoughF (ContT r m) = TypeError ... -- I believe... need to check
```

Reply:
I don't think we need a type family to do this. Could we not instead just have:

```
class PrimMonad m => PrimAffineMonad m where -- no methods
instance PrimAffineMonad m => PrimAffineMonad (MaybeT m)
instance PrimAffineMonad m => PrimAffineMonad (StateT s m)
... etc.
```

I don't see what advantage the type-family-based approach has over this.

Reply:
I just don't want to build any unnecessary evidence at runtime, but you may be right; your way may just make a pile of erased coercions. If your way makes good Core, I agree it's more pleasant.